### PR TITLE
Abort federation requests if the client disconnects early

### PR DESCRIPTION
**Pull request by richvdh**
For inbound federation requests, if a given remote server makes too many
requests at once, we start stacking them up rather than processing them
immediatedly.

However, that means that there is a fair chance that the requesting server will
disconnect before we start processing the request. In that case, if it was a
read-only request (ie, a GET request), there is absolutely no point in
building a response (and some requests are quite expensive to handle).

Even in the case of a POST request, one of two things will happen:

 * Most likely, the requesting server will retry the request and we'll get the
   information anyway.

 * Even if it doesn't, the requesting server has to assume that we didn't get
   the memo, and act accordingly.

In short, we're better off aborting the request at this point rather than
ploughing on with what might be a quite expensive request.

### CODE REVIEW

File: synapse/federation/transport/server.py [hunk 1]
```diff
@@ -338,6 +338,12 @@ async def new_func(request, *args, **kwargs):
                 if origin:
                     with ratelimiter.ratelimit(origin) as d:
                         await d
+                        if request._disconnected:
+                            logger.warning(
+                                "client disconnected before we started processing "
+                                "request"
+                            )
+                            return -1, None
```
Comment (clokep) [inline synapse/federation/transport/server.py:346]: -1 just because this is a garbage response and it doesn't really matter?

This seems similar to #7927 where `200, {}` was returned. Why the difference?

Comment (richvdh) [reply synapse/federation/transport/server.py:346]: errr good question.

`None` ensures that we don't call `respond_with_json`.

Comment (clokep) [reply synapse/federation/transport/server.py:346]: Should we be using it in the other place too then? 😄 Either way I think this is OK!

Comment (richvdh) [reply synapse/federation/transport/server.py:346]: yes, probably. I think this this the better way. we could clean up #7927 for consistency but I don't think it's a big deal really.
